src/decoder/mod.rs

```diff
@@ -1389,93 +1157,111 @@ impl<R: Read + Seek> Decoder<R> {
+    /// strip in the image as the data read may be less than the expected strip size. This interface
+    /// may change at a future date to reflect this.
+    ///
+    /// `read_strip_at` can be used to get both the strip data and the strip size.
     pub fn read_strip(&mut self) -> TiffResult<DecodingResult> {
-        self.check_chunk_type(ChunkType::Strip)?;
-        let index = self.current_chunk;
+        let result = self.read_strip_at(self.current_chunk)?;
 
-        let rows_per_strip =
-            usize::try_from(self.image().strip_decoder.as_ref().unwrap().rows_per_strip)?;
+        Ok(result.0)
+    }
 
-        let strip_height = cmp::min(
-            rows_per_strip,
-            usize::try_from(self.image().height)? - index * rows_per_strip,
-        );
+    /// Read the specified strip (at index `strip_index`) and return the binary data as a Vector and
+    /// `ChunkInfo` describing the size of the strip.
+    pub fn read_strip_at(&mut self, strip_index: usize) -> TiffResult<(DecodingResult, ChunkInfo)> {
+        self.check_chunk_type(ChunkType::Strip)?;
 
-        let mut result = self.result_buffer(usize::try_from(self.image().width)?, strip_height)?;
+        let chunk_info = self
+            .image()
+            .chunk_info(strip_index)
+            .ok_or(TiffFormatError::InvalidChunkIndex(strip_index))?;
+
+        let mut result = self.result_buffer(chunk_info.data_width, chunk_info.data_height)?;
         self.read_strip_to_buffer(result.as_buffer(0))?;
 
-        Ok(result)
+        Ok((result, chunk_info))
     }
 
-    /// Read a single tile from the image and return it as a Vector
+    /// Read a single tile from the image and return it as a Vector. This method does not return
+    /// information on the size of the tile read and so the caller must take care of tiles on the
+    /// rightmost column and the bottommost row in the image as the data read may be less than the
+    /// expected tile size. This interface may change at a future date to reflect this.
+    ///
+    /// `read_tile_at` can be used to get both the tile data and the tile size.
     pub fn read_tile(&mut self) -> TiffResult<DecodingResult> {
-        self.check_chunk_type(ChunkType::Tile)?;
+        let result = self.read_tile_at(self.current_chunk)?;
 
-        let tile = self.current_chunk;
+        self.current_chunk += 1;
 
-        let tile_attrs = self.image().tile_attributes.as_ref().unwrap();
-        let (padding_right, padding_down) = tile_attrs.get_padding(tile);
+        Ok(result.0)
+    }
 
-        let tile_width = tile_attrs.tile_width - padding_right;
-        let tile_length = tile_attrs.tile_length - padding_down;
+    /// Read the specified tile (at index `tile_index`) and return the binary data as a Vector and
+    /// `ChunkInfo` describing the size of the tile.
+    pub fn read_tile_at(&mut self, tile_index: usize) -> TiffResult<(DecodingResult, ChunkInfo)> {
```

Comment:
Could we change this function to `read_chunk_at` and get rid of `read_strip_at` above?